locator/network_topology_strategy.cc

```diff
@@ -321,17 +321,36 @@ static unsigned calculate_initial_tablets_from_topology(const schema& s, token_m
         }
         unsigned shards_in_dc = shards_per_dc_map[dc];
         unsigned tablets_in_dc = (shards_in_dc + rf_in_dc - 1) / rf_in_dc;
+        if (min_per_shard_tablet_count) {
+            auto min_tablets_in_dc = std::ceil((double)(min_per_shard_tablet_count * shards_in_dc) / rf_in_dc);
+            tablets_in_dc = std::max<unsigned>(tablets_in_dc, min_tablets_in_dc);
+        }
         initial_tablets = std::max(initial_tablets, tablets_in_dc);
     }
     rslogger.debug("Estimated {} initial tablets for table {}.{}", initial_tablets, s.ks_name(), s.cf_name());
     return initial_tablets;
 }
 
-future<tablet_map> network_topology_strategy::allocate_tablets_for_new_table(schema_ptr s, token_metadata_ptr tm, unsigned initial_scale) const {
-    auto tablet_count = get_initial_tablets();
+size_t network_topology_strategy::calculate_min_tablet_count(schema_ptr s, token_metadata_ptr tm, uint64_t target_tablet_size, std::optional<unsigned> initial_scale) const {
+    size_t tablet_count = get_initial_tablets();
+    const auto& tablet_hints = s->tablet_hints();
+    if (tablet_hints.min_tablet_count) {
+        tablet_count = std::max<size_t>(tablet_count, tablet_hints.min_tablet_count.value());
+    }
+    if (tablet_hints.expected_data_size_in_gb) {
+        tablet_count = std::max<size_t>(tablet_count, (tablet_hints.expected_data_size_in_gb.value() << 30) / target_tablet_size);
+    }
+    if (tablet_hints.min_per_shard_tablet_count) {
+        tablet_count = std::max<size_t>(tablet_count, calculate_initial_tablets_from_topology(*s, tm, _dc_rep_factor, tablet_hints.min_per_shard_tablet_count.value()));
+    }
     if (tablet_count == 0) {
-        tablet_count = calculate_initial_tablets_from_topology(*s, tm, _dc_rep_factor) * initial_scale;
+        tablet_count = calculate_initial_tablets_from_topology(*s, tm, _dc_rep_factor) * initial_scale.value_or(1);
     }
```

Comment:
it's possible for user to specify tablet_count of 0, feed hints, and expect automatic calculation to still take hints into account, no? so shouldn't we still take the maximum above when tablet_count equals 0?

Reply:
> it's possible for user to specify tablet_count of 0

You mean as keyspace `initial` tablet count?
This is processed on the first statement: `tablet_count = get_initial_tablets()`

> feed hints, and expect automatic calculation to still take hints into account, no? so shouldn't we still take the maximum above when tablet_count equals 0?

We do. `tablet_count` in this expression is the automatic variable in this function which is the maximum of all conditions.
This particular block calls `calculate_initial_tablets_from_topology`, which is the automatic calculation in the absence of other hints (initial tablets on keyspace and per-table hints)

Reply:
Ah I see now that we use tablet_count , not initial_tablets as in the keyspace. Sorry for the noise.